rust/main/utils/run-locally/src/main.rs

```diff
@@ -520,29 +459,178 @@ fn main() -> ExitCode {
                 break;
             }
         }
+        sleep(Duration::from_secs(5));
+    }
+
+    if failure_occurred {
+        return report_test_result(failure_occurred);
+    }
+
+    // Here we want to restart the relayer and validate
+    // its restart behaviour.
+    restart_relayer(&config, &mut state, &rocks_db_dir);
+
+    // give relayer a chance to fully restart.
+    sleep(Duration::from_secs(20));
+
+    let loop_start = Instant::now();
 
-        // verify long-running tasks are still running
-        for (name, (child, _)) in state.agents.iter_mut() {
-            if let Some(status) = child.try_wait().unwrap() {
-                if !status.success() {
-                    log!(
-                        "Child process {} exited unexpectedly, with code {}. Shutting down",
-                        name,
-                        status.code().unwrap()
-                    );
-                    failure_occurred = true;
-                    SHUTDOWN.store(true, Ordering::Relaxed);
-                    break;
-                }
+    while !SHUTDOWN.load(Ordering::Relaxed) {
+        if config.ci_mode {
+            if relayer_restart_invariants_met().unwrap_or(false) {
+                // end condition reached successfully
+                break;
+            } else if (Instant::now() - loop_start).as_secs() > config.ci_mode_timeout {
+                // we ran out of time
+                log!("CI timeout reached before relayer restart invariants were met");
+                failure_occurred = true;
+                break;
             }
         }
```

Comment:
this while loop follows a very similar pattern to the one above and could be refactored

Reply:
Refactored into function `wait_for_condition`